### PR TITLE
support flow description YAML

### LINKED ISSUE #505: support flow description

### PR DESCRIPTION
**Pull request by gituser173**
Fixes #505
 
Signed-off-by: Claudiu Ban claudiu-laurian.ban@hpe.com

### CODE REVIEW

**Comment by gituser173** [comment]
@Bonczidai please review(I will add the unit tests after your review)

File: cloudslang-all/src/main/java/io/cloudslang/lang/api/Slang.java [hunk 1]
```diff
@@ -31,7 +32,14 @@
...
      */
-    public CompilationArtifact compile(SlangSource source, Set<SlangSource> dependencies);
+    CompilationArtifact compile(SlangSource source, Set<SlangSource> dependencies);
+
+    /**
+     * Extract the metadata of a flow or operation written in CloudSlang
+     * @param source the CloudSlang source containing the flow
+     * @return the metadata of the flow
+     */
+    Metadata extractMetadata(SlangSource source);
```
Comment (Bonczidai) [inline cloudslang-all/src/main/java/io/cloudslang/lang/api/Slang.java]: I'd put this method first for logical order

File: cloudslang-cli/src/main/java/io/cloudslang/lang/cli/utils/MetadataHelperImpl.java [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+ * Date: 1/11/2016
+ */
+@Component
+public class MetadataHelperImpl implements MetadataHelper {
+
+    @Autowired
+    private Slang slang;
+
+    @Override
+    public Metadata extractMetadata(File file) throws IOException {
```
Comment (Bonczidai) [inline cloudslang-cli/src/main/java/io/cloudslang/lang/cli/utils/MetadataHelperImpl.java]: Y declare IOException here?

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/YAMLMetadataExtractorImpl.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * User: bancl
+ * Date: 1/11/2016
+ */
+@Component
+public class YAMLMetadataExtractorImpl implements MetadataExtractor {
```
Comment (Bonczidai) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/YAMLMetadataExtractorImpl.java]: according to convention, class name should be MetadataExtractorImpl

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/YAMLMetadataExtractorImpl.java [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        return inputs;
+    }
+
+    private String extractDescriptionYAMLString(SlangSource source) {
+        StringBuilder sb = new StringBuilder();
+        try (BufferedReader reader = new BufferedReader(new StringReader(source.getSource()))) {
+            String line = reader.readLine();
+            while (line != null) {
+                if (line.startsWith("##")) {
+                    line = line.substring(3);
```
Comment (Bonczidai) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/YAMLMetadataExtractorImpl.java]: does it work for such line:
```
##
```
?

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/YAMLMetadataExtractorImpl.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+                    if (!line.startsWith("#"))
+                        sb.append(line).append(System.getProperty("line.separator"));
+                }
+                line = reader.readLine();
+            }
+        } catch (IOException e) {
+            throw new RuntimeException("Read issue(should not happen)", e);
```
Comment (Bonczidai) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/YAMLMetadataExtractorImpl.java]: error message should be more descriptive ~ Error processing metadata, error extracting metadata from <source name>

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/YAMLMetadataExtractorImpl.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+        }
+        return sb.toString();
+    }
+}
```
Comment (Bonczidai) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/YAMLMetadataExtractorImpl.java]: you extract `##`, `3` and exception message as consts

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/model/Metadata.java [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    private Map<String, String> inputs;
+    private Map<String, String> outputs;
+    private Map<String, String> results;
+    private String description;
+
+    public Map<String, String> getInputs() {
+        return inputs;
+    }
+
+    public void setInputs(Map<String, String> inputs) {
```
Comment (Bonczidai) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/model/Metadata.java]: is there any reason for setter here? better add a constructor with params

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/model/MetadataYAML.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+
+    public void setInputs(List<Map<String, String>> inputs) {
+        this.inputs = inputs;
+    }
+
+    public List<Map<String, String>> getOutputs() {
+        return outputs;
+    }
+
+    public void setOutputs(List<Map<String, String>> outputs) {
```
Comment (Bonczidai) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/model/MetadataYAML.java]: same for setters here

**Comment by Bonczidai** [comment]
little redesign suggestion in compiler package:

it has three phases: parse (you get the doc as maps, lists, strings etc.) -> model (you get objects of your choice) -> compile to Execution Plan (not affected here)

- move `MetadatYAML` to `parser.model` - rename ~ `ParsedMetadata`
- `Metadata` stays in `modeller.model`
- break `MetadataExtractor` in three:
  - `MetadataParser` returns `MetadatYAML` in `parser` package
  - `MetadataModeller` returns `Metadata` in `modeller.model` package
  - `MetadataExtractor` wires the two and exposes the interface 

WDYT?

**Comment by Bonczidai** [comment]
:ok_hand:  good work

please handle comments and add tests

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/model/MetadataYAML.java (hunk 8 above)
Comment (gituser173) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/model/MetadataYAML.java]: the snakeyaml library doesn't seem to work when I remove the setters

File: cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/model/Metadata.java (hunk 7 above)
Comment (gituser173) [inline cloudslang-compiler/src/main/java/io/cloudslang/lang/compiler/modeller/model/Metadata.java]: I find it easier to read separate setters than a constructor/method with many parameters

**Comment by gituser173** [comment]
about the redesign suggestions:
I would not break the MetadataExtractor in three unless we have a use for MetadataParser that returns MetadatYAML in parser package in other places. MetadataModeler is just a converter from MetadataYAML to Metadata pojos, don't think it would be of any use anywhere else, I would keep those two small private methods inside MetadataExtractor. 

Moved and renamed MetadataYAML to ParsedMetada :)

**Comment by gituser173** [comment]
@SamMarkowitz 
capture

**Comment by SamMarkowitz** [comment]
Hmm. I think all the bracketing, commas, and quotes of what I assume is a JSON object is a bit confusing. Can all that be stripped away?

Also, there are "\n"s where in the documentation there are new lines.

**Comment by gituser173** [comment]
@SamMarkowitz 
capture

**Comment by Bonczidai** [comment]
two Qs:
- why using `.` separator in file names instead of `_`, we should follow our conventions IMHO
- how does the CLI output look like when inspecting an invalid description - please put a screenshot here

**Comment by SamMarkowitz** [comment]
Looks great! For the inputs can we get the dash after the pipe to show up on the next line or get rid of the dashes altogether (which I think is the last way we had it in issue 505).

```
inputs:
  json_input:
    JSON data input
    example: some example   
```

**Comment by gituser173** [comment]
@Bonczidai 
capture

**Pull request merged**